garminexport/garminbackup.py

```diff
@@ -0,0 +1,129 @@
+    parser.add_argument(
+        "-E", "--ignore-errors", action='store_true',
+        help="Ignore errors and keep going. Default: FALSE")
+    parser.add_argument(
+        "--max-retries", metavar="NUM", default=DEFAULT_MAX_RETRIES,
+        type=int, help="The maximum number of retries to make on failed attempts to fetch an activity. Exponential backoff will be used, meaning that the delay between successive attempts will double with every retry, starting at one second. DEFAULT: %d" % DEFAULT_MAX_RETRIES)
+
+    return parser.parse_args()
+
+
+def garminbackup(args=None):
+
+    if not args.log_level in LOG_LEVELS:
+        raise ValueError("Illegal log-level: {}".format(args.log_level))
+
+    # if no --format was specified, all formats are to be backed up
+    args.format = args.format if args.format else export_formats
+    log.info("backing up formats: %s", ", ".join(args.format))
+
+    logging.root.setLevel(LOG_LEVELS[args.log_level])
+
+    try:
+        if not os.path.isdir(args.backup_dir):
+            os.makedirs(args.backup_dir)
+
+        if not args.password:
+            args.password = getpass.getpass("Enter password: ")
+
+        # set up a retryer that will handle retries of failed activity
+        # downloads
+        retryer = Retryer(
+            delay_strategy=ExponentialBackoffDelayStrategy(
+                initial_delay=timedelta(seconds=1)),
+            stop_strategy=MaxRetriesStopStrategy(args.max_retries))
+
+
+        with GarminClient(args.username, args.password) as client:
+            # get all activity ids and timestamps from Garmin account
+            log.info("scanning activities for %s ...", args.username)
+            activities = set(retryer.call(client.list_activities))
+            log.info("account has a total of %d activities", len(activities))
+
+            missing_activities = garminexport.backup.need_backup(
+                activities, args.backup_dir, args.format)
+            backed_up = activities - missing_activities
+            log.info("%s contains %d backed up activities",
+                args.backup_dir, len(backed_up))
+
+            log.info("activities that aren't backed up: %d",
+                     len(missing_activities))
+
+            for index, activity in enumerate(missing_activities):
+                id, start = activity
+                log.info("backing up activity %d from %s (%d out of %d) ..." % (id, start, index+1, len(missing_activities)))
+                try:
+                    garminexport.backup.download(
+                        client, activity, retryer, args.backup_dir,
+                        args.format)
+                except Exception as e:
+                    log.error(u"failed with exception: %s", e)
+                    if not args.ignore_errors:
+                        raise
+    except Exception as e:
+        exc_type, exc_value, exc_traceback = sys.exc_info()
+        log.error(u"failed with exception: %s", str(e))
```

Comment:
Any errors should probably be propagated all the way to the initial caller. This exception handling belongs in "main".

Reply:
moved to the root script. btw, what purpose serves this line: ```exc_type, exc_value, exc_traceback = sys.exc_info()```? returned objects do not seem to be used anywhere...

Reply:
Must be a leftover from an older version of the code. Feel free to clean up! :)

Reply:
removed the line